### ISSUE TITLE
Implement Batch-Normalization

### ISSUE THREAD

**Issue by lucasb-eyer** [lucasb-eyer/DeepFried#1]
http://arxiv.org/abs/1502.03167

**Comment by lucasb-eyer** [comment]
So turns out this also makes a difference between training-time and testing-time, just like Dropout #2 . Guess I'll have to refactor a bit so I can easily implement both of them.

**Comment by lucasb-eyer** [comment]
The refactor is happening simultaneously to the export =)

**Comment by lucasb-eyer** [comment]
Turns out this is harder than that, and not "just like Dropout." Ideally, it'd work in the following way:
1. Train the network, nothing special
2. Build the prediction network by:
   1. Sending all training data through the trained network up to the first BN-layer.
   2. Compute the training's mean and variance here.
   3. Use these to compute this BN's prediction γ and β as in Line 11 of the paper.
   4. Sending all training data through the trained network up to the second BN-layer.
   5. ...

As an easy stop-gap, one can use the prediction's minibatch statistics. People [report that working well](https://github.com/benanne/Lasagne/issues/141#issuecomment-75436361), it just makes prediction nondeterministic.
